### ISSUE TITLE
Unification of model initialization methods / naming across domain libraries

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#65153]
E.g. torchvision often calls it `_init_weights` ,  https://github.com/pytorch/vision/issues/3410

New guidance is also using nn.init methods for supporting skipinig instead of direct use of `normal_` and `fill_`:
https://pytorch.org/tutorials/prototype/skip_param_init.html#updating-modules-to-support-skipping-initialization

Right now it's often done ad-hoc.

**Comment by vadimkantorov** [comment]
Maybe related:https://github.com/pytorch/pytorch/issues/29523

**Comment by albanD** [comment]
If you're looking for solution that usually work (even though not guaranteed), they are detailed in https://github.com/huggingface/transformers/issues/21913#issuecomment-1453858274 and that will be more reliable than patching nn.init

**Comment by vadimkantorov** [comment]
I think what's needed is meta-device usage + skipping nn.init + lazy loading from load_state_dict (https://github.com/pytorch/pytorch/issues/79967) combined with contig placing/materialization on the correct device. Most or all of the tricks are already there...

This issue is also about coming to peace that reset_parameters methods are useful (even if only for dynamic execution) and should not be inlined for the fear that they would not work correctly after FX-tracing (if they are needed to be workable under tracing, some proxy objects could be created so that getattr of weight parameters gets redirected after FX-tracing). Even if they cannot be propagated recursively always consistently, they are still a useful idiom to have on common modules like Convs / Linear / ResNet blocks etc - much better than people having to be looking for the source code and copy-pasting it if needing to reset some model heads.

**Comment by albanD** [comment]
> This issue is also about coming to peace that reset_parameters methods are useful

I definitely agree with that! My stance on this is that there is no way for us to enforce that as a standard on nn.Module. And even if we do, it will take ~years for it to be spread wide enough that we can start relying on their existence.

**Comment by vadimkantorov** [comment]
I think:
- they already exist across the domain libraries owned by pytorch, so renaming them there to be the same idiom is already good
- even without it being used everywhere, it's already useful. and people tend to adopt idioms from pytorch repos if the idioms/api contracts seem well designed. so in case someone writes-up a `_init_weights` method, they would be happily adopt semi-standard idiomatic naming, if it's suggested in tutorials / well-spread in the official libraries / present in docs even as "recommended" modality
- so kind of "relying on their existence" aspect is a bit irrelevant, as they are still a bit niche. after they are implemented in at least some useful parts (e.g. convs/linear/attention blocks), more idioms can be developed

**Comment by vadimkantorov** [comment]
@albanD btw, is `with torch.device('meta')` supposed to work? then maybe `skip_init` could use it instead of passing explicit `device = 'meta'` (although I like a lot support for passing some options explicitly without relying on magic context managers :))

**Comment by albanD** [comment]
The problem is that `with torch.device('meta')` will not always work and can fail in arbitrary ways when it doesn;t. From a nice error like https://github.com/pytorch/pytorch/issues/96409 to things silently not being on the device (if device is explicitly provided by the user) or tripping user code (if user assumed Tensor is on CPU).
The reason we went with the explicit API here is that it is guaranteed to either work as the Module writer intended or raise a clear error (device arg not supported for a given Module).

I don't think we should offer APIs that can fail in arbitrary ways as torch.nn APIs. That is why `torch.device()` is not used there.

**Comment by vadimkantorov** [comment]
Also, weight norm should probably be moved to parametrization api :) Maybe this would fix the meta device problem with it?

But overall, I think there should first be a table on various models supporting explicit device or with torch.device('meta') for this purpose. My understanding is that currently HuggingFace still figures it out but with very ugly hacks (storage swaps etc). It would be good if these hacks were upcycled to something nicer / idiomatic (even if not 100% supported for legacy codebases)

**Comment by albanD** [comment]
> Also, weight norm should probably be moved to parametrization api :) Maybe this would fix the meta device problem with it?

Yes it should! There is an issue about it somewhere that should be "actionable"!
